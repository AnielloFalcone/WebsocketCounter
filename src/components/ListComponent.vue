<template>
    <div id="list" v-if="orders.length > 0">
        <list-item-component v-for="order in orders"
                             v-bind:key="order.id"
                             v-bind:order="order"
        ></list-item-component>
    </div>
</template>

<script>
    import ListItemComponent from "./ListItemComponent";

    export default {
        name: "ListComponent",
        components: {ListItemComponent},
        props: ['counter'],
        data() {
            return {
                orders: [
                    {
                        numberLabel: 'Num',
                        idLabel: 'Order id',
                        amountLabel: 'Amount'
                    }
                ]
            }
        },
        created() {
            this.createOrders();
        },
        watch: {
            counter: {
                handler(newCounter, oldCounter) {
                    if (newCounter - oldCounter === 1) {
                        this.setOrder(newCounter, this.scrollToLastOrder);
                    }
                    else {
                        const min = oldCounter + 1;
                        this.createOrders(min, newCounter, this.scrollToLastOrder);
                    }
                }
            }
        },
        methods: {
            createOrders(min = 1, max = this.counter, scrollCallback = undefined) {
                for (let i = min; i <= max; i++) {
                    this.setOrder(i);
                }

                if (scrollCallback) {
                    scrollCallback();
                }
            },
            getUuidv4() {
                return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
                    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
                );
            },
            setOrder(number, scrollCallback = undefined) {
                this.orders.push({
                    amount: Math.floor(Math.random() * 1000) + 1,
                    id: this.getUuidv4(),
                    number,
                });

                if (scrollCallback) {
                    scrollCallback();
                }
            },
            scrollToLastOrder() {
                setTimeout(() => {
                    window.scrollTo(0, document.body.scrollHeight);
                });
            }
        }
    }
</script>

<style scoped>
    #list {
        width: 60%;
    }
</style>
