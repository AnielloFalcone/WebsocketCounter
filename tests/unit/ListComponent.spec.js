import { shallowMount } from '@vue/test-utils';
import ListComponent from '../../src/components/ListComponent';

describe('ListComponent.vue', () => {
    const counter = 6;
    let wrapper;

    beforeAll(() => {
        const crypto = require('crypto');

        Object.defineProperty(global.self, 'crypto', {
            value: {
                getRandomValues: arr => crypto.randomBytes(arr.length),
            },
        });

    });

    beforeEach(() => {
        wrapper = shallowMount(ListComponent, {
            propsData: {
                counter
            },
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders ListComponent when passed', () => {
        expect(wrapper.isVueInstance()).toBeTruthy();
        expect(wrapper.element).toBeDefined();
        expect(wrapper.element).toMatchSnapshot();
    });

    it('props should be defined', () => {
        expect(wrapper.props().counter).toBe(counter);
    });

    /**
     * Expected length + 1 because of the table header
     */
    it('orders should defined and have same counter length + 1', () => {
        expect(wrapper.vm.orders).toHaveLength(counter + 1);
    });
});
