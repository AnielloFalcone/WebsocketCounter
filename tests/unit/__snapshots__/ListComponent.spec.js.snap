// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent.vue renders ListComponent when passed 1`] = `
<div
  id="list"
>
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
  <list-item-component-stub
    order="[object Object]"
  />
</div>
`;
